internal/storage/sql/common/namespace.go

```diff
@@ -0,0 +1,225 @@
+package common
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	sq "github.com/Masterminds/squirrel"
+	errs "go.flipt.io/flipt/errors"
+	"go.flipt.io/flipt/internal/storage"
+	fliptsql "go.flipt.io/flipt/internal/storage/sql"
+	flipt "go.flipt.io/flipt/rpc/flipt"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+func (s *Store) GetNamespace(ctx context.Context, key string) (*flipt.Namespace, error) {
+	var (
+		createdAt fliptsql.Timestamp
+		updatedAt fliptsql.Timestamp
+
+		namespace = &flipt.Namespace{}
+
+		err = s.builder.Select("\"key\", name, description, protected, created_at, updated_at").
+			From("namespaces").
+			Where(sq.Eq{"key": key}).
+			QueryRowContext(ctx).
+			Scan(
+				&namespace.Key,
+				&namespace.Name,
+				&namespace.Description,
+				&namespace.Protected,
+				&createdAt,
+				&updatedAt)
+	)
+
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, errs.ErrNotFoundf(`namespace "%s"`, key)
+		}
+
+		return nil, err
+	}
+
+	namespace.CreatedAt = createdAt.Timestamp
+	namespace.UpdatedAt = updatedAt.Timestamp
+
+	return namespace, nil
+}
+
+func (s *Store) ListNamespaces(ctx context.Context, opts ...storage.QueryOption) (storage.ResultSet[*flipt.Namespace], error) {
+	params := &storage.QueryParams{}
+
+	for _, opt := range opts {
+		opt(params)
+	}
+
+	var (
+		namespaces []*flipt.Namespace
+		results    = storage.ResultSet[*flipt.Namespace]{}
```

Comment:
Just sharing this thought out loud. Not needed for this PR.

Now that we have the generic `ResultSet` and we have `QueryParams` we could definitly create a generic paginated query function.
It would likely need to take:
- an initial query builder (containing the select, from and so on)
- a generic function which can scan a single row into the generic type

Then it could do the adding of order, limit (+1), page token parsing, offset, calculating of next token and returning of generic result set.

Food for thought :D

Reply:
yeah 💯 i was thinking about this as I was going through it. I'll work on another PR for creating the generic func